File: packages/core/src/markdown.ts
/**
 * Telegram MarkdownV2 escaping + scrubbing for prompt-injection vectors.
 * https://core.telegram.org/bots/api#markdownv2-style
 */
const SPECIAL = /([_*\[\]()~`>#+\-=|{}.!\\])/g;

export function escapeMdV2(text: string): string {
  return text.replace(SPECIAL, "\\$1");
}

/** Wrap an untrusted (fetched / user-provided) string for LLM input. */
export function untrusted(label: string, text: string): string {
  const safe = text
    .replace(/<\/?untrusted_source[^>]*>/gi, "[redacted-tag]")
    .slice(0, 32_000);
  return `<untrusted_source name="${label}">\n${safe}\n</untrusted_source>`;
}

/** Strip suspicious markdown links + zero-width chars before showing to LLM. */
export function scrubForModel(text: string): string {
  return text
    .replace(/[​-‏‪-‮⁠﻿]/g, "")
    .replace(/!\[[^\]]*]\([^)]+\)/g, "[image]")
    .slice(0, 32_000);
}

export type DigestPost = {
  title: string;
  source: string;
  topic: string;
  score: number;
  tldr: string;
  bullets: string[];
  url: string;
};

const ZeroWidth = /[​-‏‪-‮⁠﻿]/g;

function clean(s: string): string {
  return s.replace(ZeroWidth, "").trim();
}

export function formatItemMessage(p: DigestPost): string {
  const title = escapeMdV2(clean(p.title));
  const source = escapeMdV2(clean(p.source));
  const topic = escapeMdV2(clean(p.topic));
  const score = escapeMdV2(p.score.toFixed(0));
  const tldr = escapeMdV2(clean(p.tldr));
  const bullets = p.bullets
    .slice(0, 3)
    .map((b) => `• ${escapeMdV2(clean(b))}`)
    .join("\n");
  // URL inside () must have ) and \ escaped
  const url = p.url.replace(/[\\)]/g, (c) => `\\${c}`);
  return [
    `*${title}*`,
    `_${source} · ${topic} · score ${score}/10_`,
    "",
    tldr,
    "",
    bullets,
    "",
    `🔗 [link](${url})`,
  ].join("\n");
}

File: core/markdown.py
import re

_SPECIAL = re.compile(r"([_*\[\]()~`>#+\-=|{}.!\\])")
_ZERO_WIDTH = re.compile("[\u200b-\u200f\u202a-\u202e\u2060\ufeff]")
_IMAGE_LINK = re.compile(r"!\[[^\]]*]\([^)]+\)")


def escape_md_v2(text: str) -> str:
    return _SPECIAL.sub(r"\\\1", text)


def untrusted(label: str, text: str) -> str:
    safe = re.sub(
        r"</?untrusted_source[^>]*>", "[redacted-tag]", text, flags=re.IGNORECASE
    )
    safe = safe[:32_000]
    return f'<untrusted_source name="{label}">\n{safe}\n</untrusted_source>'


def scrub_for_model(text: str) -> str:
    text = _ZERO_WIDTH.sub("", text)
    text = _IMAGE_LINK.sub("[image]", text)
    # Defence-in-depth: strip injection tags before untrusted() wraps them.
    text = re.sub(
        r"</?untrusted_source[^>]*>", "[redacted-tag]", text, flags=re.IGNORECASE
    )
    return text[:32_000]


def _clean(s: str) -> str:
    return _ZERO_WIDTH.sub("", s).strip()


def format_item_message(
    title: str,
    source: str,
    topic: str,
    score: float,
    tldr: str,
    bullets: list[str],
    url: str,
) -> str:
    t = escape_md_v2(_clean(title))
    src = escape_md_v2(_clean(source))
    tpc = escape_md_v2(_clean(topic))
    sc = escape_md_v2(str(round(score)))
    tl = escape_md_v2(_clean(tldr))
    bullet_lines = "\n".join(f"• {escape_md_v2(_clean(b))}" for b in bullets[:3])
    # URL inside () must have ) and \ escaped
    safe_url = re.sub(r"[\\)]", lambda m: f"\\{m.group()}", url)
    return "\n".join(
        [
            f"*{t}*",
            f"_{src} · {tpc} · score {sc}/10_",
            "",
            tl,
            "",
            bullet_lines,
            "",
            f"🔗 [link]({safe_url})",
        ]
    )
